robot_skills/src/robot_skills/perception.py

```diff
@@ -296,31 +308,22 @@ def get_rgb_depth_caminfo(self, timeout=5):
         :param timeout: How long to wait until the images are all collected.
         :return: tuple(rgb, depth, depth_info) or a None if no images could be gathered.
         """
-        event = Event()
-
-        def callback(rgb, depth, depth_info):
-            rospy.loginfo('Received rgb, depth, cam_info')
-            self._image_data = (rgb, depth, depth_info)
-            event.set()
-
-        # camera topics
-        depth_info_sub = message_filters.Subscriber('{}/depth_registered/camera_info'.format(self._camera_base_ns),
-                                                    CameraInfo)
-        depth_sub = message_filters.Subscriber('{}/depth_registered/image'.format(self._camera_base_ns), Image)
-        rgb_sub = message_filters.Subscriber('{}/rgb/image_raw'.format(self._camera_base_ns), Image)
-
-        ts = message_filters.ApproximateTimeSynchronizer([rgb_sub, depth_sub, depth_info_sub],
-                                                         queue_size=1,
-                                                         slop=10)
-        ts.registerCallback(callback)
-        event.wait(timeout)
-        ts.callbacks.clear()
-        del ts, depth_info_sub, depth_sub, rgb_sub, callback
+        if self._depth_sub is None:
+            self._depth_info_sub = message_filters.Subscriber(
+                '{}/depth_registered/camera_info'.format(self._camera_base_ns),
+                CameraInfo)
+            self._depth_sub = message_filters.Subscriber('{}/depth_registered/image'.format(self._camera_base_ns), Image)
+            self._rgb_sub = message_filters.Subscriber('{}/rgb/image_raw'.format(self._camera_base_ns), Image)
+
+            self._ts = message_filters.ApproximateTimeSynchronizer([self._rgb_sub, self._depth_sub, self._depth_info_sub],
+                                                                   queue_size=1, slop=10)
+            self._ts.registerCallback(self._sync_callback)
+            rospy.sleep(2.0)
 
         if any(self._image_data):
             return self._image_data
         else:
-            return None
+            return (None, None, None)
```

Comment:
This way the return value is always a tuple. So you can directly unpack the tuple.

Reply:
That is ease of notation but it will probably lead to people neglecting to check for None. I think None is a better indicator that no result could be found rather than a tuple of Nones. you indeed lose the directly unpacking but I don't mind that. Its a few extra lines of code.

Reply:
@ar13pit: Two tuples are more consistent than a tuple and a None. So less buggy code